File: sample-repos/acme-platform-go/pkg/auth/middleware.go
package auth

import (
	"context"
	"fmt"
	"net/http"
	"strings"
)

// AcmeAuthConfig configures the Acme authentication middleware.
type AcmeAuthConfig struct {
	AuthServiceURL string // Default: "https://auth.internal.acme.com"
	RequiredScopes []string
	AllowInternal  bool // Allow X-Acme-Internal-Service header
}

type contextKey string

const (
	userContextKey  contextKey = "acme_user"
	headerAcmeAuth  string     = "X-Acme-Auth"
	headerInternal  string     = "X-Acme-Internal-Service"
)

// AcmeUser represents a validated user from the Acme auth service.
type AcmeUser struct {
	ID         string
	Email      string
	TeamLabel  string
	Scopes     []string
	IsInternal bool
}

// NewAuthMiddleware creates Acme-standard authentication middleware.
// All HTTP services MUST use this instead of rolling custom auth.
// Error codes: ACME-AUTH-001 (missing token), ACME-AUTH-002 (invalid token),
// ACME-AUTH-003 (insufficient scopes)
func NewAuthMiddleware(cfg AcmeAuthConfig) func(http.Handler) http.Handler {
	if cfg.AuthServiceURL == "" {
		cfg.AuthServiceURL = "https://auth.internal.acme.com"
	}

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Check for internal service-to-service calls
			if cfg.AllowInternal {
				if svc := r.Header.Get(headerInternal); svc != "" {
					user := &AcmeUser{
						ID:         fmt.Sprintf("svc:%s", svc),
						IsInternal: true,
					}
					ctx := context.WithValue(r.Context(), userContextKey, user)
					next.ServeHTTP(w, r.WithContext(ctx))
					return
				}
			}

			token := r.Header.Get(headerAcmeAuth)
			if token == "" {
				http.Error(w, `{"error":"ACME-AUTH-001","message":"missing X-Acme-Auth header"}`, http.StatusUnauthorized)
				return
			}

			user, err := validateToken(cfg.AuthServiceURL, token)
			if err != nil {
				http.Error(w, `{"error":"ACME-AUTH-002","message":"invalid token"}`, http.StatusUnauthorized)
				return
			}

			if !hasRequiredScopes(user.Scopes, cfg.RequiredScopes) {
				http.Error(w, `{"error":"ACME-AUTH-003","message":"insufficient scopes"}`, http.StatusForbidden)
				return
			}

			ctx := context.WithValue(r.Context(), userContextKey, user)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

// GetUser extracts the authenticated Acme user from the request context.
func GetUser(ctx context.Context) (*AcmeUser, bool) {
	user, ok := ctx.Value(userContextKey).(*AcmeUser)
	return user, ok
}

func validateToken(authURL, token string) (*AcmeUser, error) {
	// In production, calls auth.internal.acme.com/v1/validate
	return &AcmeUser{ID: "user-123", Email: "dev@acme.com"}, nil
}

func hasRequiredScopes(userScopes, required []string) bool {
	scopeSet := make(map[string]bool)
	for _, s := range userScopes {
		scopeSet[s] = true
	}
	for _, r := range required {
		if !scopeSet[r] {
			return false
		}
	}
	return true
}

File: sample-repos/acme-platform-python/acme/auth/middleware.py
"""
Acme Authentication Middleware for Flask.

Validates X-Acme-Auth header on every request.
Error codes: ACME-AUTH-001 (missing), ACME-AUTH-002 (invalid), ACME-AUTH-003 (insufficient scopes).
"""

import functools
from flask import request, jsonify, g


class AcmeUser:
    """Represents a validated user from the Acme auth service."""

    def __init__(self, user_id: str, email: str, team_label: str, scopes: list = None, is_internal: bool = False):
        self.user_id = user_id
        self.email = email
        self.team_label = team_label
        self.scopes = scopes or []
        self.is_internal = is_internal


def acme_auth_required(scopes: list = None, allow_internal: bool = False):
    """Decorator for Acme-standard authentication on Flask routes.

    All HTTP endpoints MUST use this instead of rolling custom auth.

    Usage:
        @app.route("/api/data")
        @acme_auth_required(scopes=["read:data"])
        def get_data():
            user = g.acme_user
            ...
    """
    def decorator(f):
        @functools.wraps(f)
        def wrapper(*args, **kwargs):
            # Check for internal service-to-service calls
            if allow_internal:
                internal_service = request.headers.get("X-Acme-Internal-Service")
                if internal_service:
                    g.acme_user = AcmeUser(
                        user_id=f"svc:{internal_service}",
                        email="",
                        team_label="",
                        is_internal=True,
                    )
                    return f(*args, **kwargs)

            token = request.headers.get("X-Acme-Auth")
            if not token:
                return jsonify({"error": "ACME-AUTH-001", "message": "missing X-Acme-Auth header"}), 401

            user = _validate_token(token)
            if not user:
                return jsonify({"error": "ACME-AUTH-002", "message": "invalid token"}), 401

            if scopes and not _has_required_scopes(user.scopes, scopes):
                return jsonify({"error": "ACME-AUTH-003", "message": "insufficient scopes"}), 403

            g.acme_user = user
            return f(*args, **kwargs)

        return wrapper
    return decorator


def _validate_token(token: str) -> AcmeUser:
    """Validate token against auth.internal.acme.com/v1/validate."""
    # In production, this calls the Acme auth service
    return AcmeUser(user_id="user-123", email="dev@acme.com", team_label="platform")


def _has_required_scopes(user_scopes: list, required: list) -> bool:
    return all(s in user_scopes for s in required)
